### PR TITLE
feature: Command line arguments - version and platform

### PR DESCRIPTION
**Pull request by haadcode**
This PR will:

Add possibility to set the wanted go-ipfs version in package.json.
Add possibility to specify the version to download.
Add possibility to specify the platform to download.
Add possibility to specify the arch to download.
Add possibility to specify the download url/domain.
New public API.
Add simple stdout output for download progress.
Catch errors and add output them to stderr.
Add stdout output for user feedback.
Add more tests.

### CODE REVIEW

**Comment by haadcode** [comment]
@diasdavid @dignifiedquire can I get a CR for this? This will add the ability to specify version and platform and the module becomes 200% more useful. This will also make it work on CI to build for any platform (eg. together with Electron), ie. I can include Windows version of go-ipfs on Linux machine.

It's starting work quite nicely as a "go-ipfs version manager", perhaps at some point we could write one on top of this.

File: package.json [hunk 1]
```diff
@@ -4,7 +4,7 @@
   "description": "Install the latest go-ipfs binary",
   "main": "src/index.js",
   "scripts": {
-    "install": "node src/bin.js",
+    "install": "node src/bin.js $TARGET_VERSION $TARGET_OS",
```
Comment (dignifiedquire) [inline package.json]: This will not work on windows, but there is https://www.npmjs.com/package/cross-env to fix that

File: src/bin.js [hunk 2]
```diff
@@ -2,4 +2,6 @@
 
 'use strict'
 
-require('./')()
```
Comment (dignifiedquire) [inline src/bin.js]: why the change of permissions?

File: src/bin.js [hunk 3]
```diff
@@ -2,4 +2,6 @@
...
+// First param is the target version
+// Second param is the target platform
+require('./')(process.argv[2], process.argv[3], function () {})
```
Comment (dignifiedquire) [inline src/bin.js]: Rather than passing in an empty function, why not make the callback either optional or useful and print something like `successfully installed <version>`

File: package.json (hunk 1 above)
Comment (haadcode) [inline package.json]: Is there any cross platform solution that doesn't require pulling in a dependency?

File: src/bin.js (hunk 2 above)
Comment (haadcode) [inline src/bin.js]: Opps, didn't need to. Will revert that change.

File: src/bin.js (hunk 3 above)
Comment (haadcode) [inline src/bin.js]: Later. Right now it works fine without.

File: package.json (hunk 1 above)
Comment (dignifiedquire) [inline package.json]: not as far as I know, which is why that package exists in the first place

Comment (haadcode) [inline package.json]: What happens on Windows with the command? Does `process.argv` get the params as strings or nothing? If latter, then I think we're fine for now, when the params are undefined, it'll download the current platform and version from the package.

If they do get passed in as strings on windows, then we need to figure out something better.

Comment (dignifiedquire) [inline package.json]: it probably throws because the envvariables are not valid syntax on windows shells

Comment (dignifiedquire) [inline package.json]: so this will literally stop any windows users from installing

File: src/bin.js (hunk 3 above)
Comment (diasdavid) [inline src/bin.js]: Can we make that 'later' more explicit? With a TODO of what is desired for example.

File: src/bin.js (hunk 2 above)
Comment (haadcode) [inline src/bin.js]: Changed back to 644

**Comment by haadcode** [comment]
I removed the env variables from the npm script to the actual version setup in index.js. This way it can pick up the platform and version from a) arguments passed in b) environment variables and c) defaults (package.json and go-platform).

Also cleaned up the callback so we don't need to discuss what we put there. And changed the file permission back to original. 

Everything addressed, can we get final CR and if all good, merge this?

File: src/index.js [hunk 4]
```diff
@@ -7,26 +7,48 @@ var path = require('path')
...
   }
 
-  const isWindows = checkPlatform.isWindows(goenv.GOOS)
+  // Set the platform either from arguments or goenv OS
+  const isWindows = checkPlatform.isWindows(platform)
 
   // hacky hack hack to work around unpublishability
-  version = version.replace(/-[0-9]+/, '')
+  // version = version.replace(/-[0-9]+/, '')
+  // What is this^^?? @haadcode
```
Comment (dignifiedquire) [inline src/index.js]: Not sure we can drop this, because the version in `package.json` corresponds to go version, but we sometimes have multiple versions of this package per go-ipfs version, so this fixes that, i.e.

| go-ipfs | package.json |
| 0.4.4    | 0.4.4    |
| 0.4.4    | 0.4.4-1 | 
| 0.4.4    | 0.4.4-2 |
| 0.4.4    | 0.4.4-3 |

File: src/index.js [hunk 5]
```diff
@@ -1,67 +1,121 @@
...
+    })
+  })
 }
 
-function noop () {}
+// Public interface
+Object.assign(download, { Platforms: support.Platforms })
+Object.assign(download, { Versions: support.Versions })
+Object.assign(download, { Archs: support.Archs })
+Object.assign(download, { Download: download })
```
Comment (dignifiedquire) [inline src/index.js]: As you are not checking for existence, you could just

```js
download.Platforms = support.Platforms,
download.Versions = support.Versions
```

also why are these captialized, they are not constructors, so I think they should be lower case

**Comment by haadcode** [comment]
Everything got refactored and updated. Added more thorough support for arguments (arch, dist url). Squashed all commits.

Ready for review!

File: Makefile [hunk 6]
```diff
@@ -0,0 +1,15 @@
+all: test
```
Comment (dignifiedquire) [inline Makefile]: I am 👎 on adding makefiles, they are not cross platform and don't add much value compared to just `npm run-script`

Comment (haadcode) [inline Makefile]: I can remove the Makefile from this PR so that we don't need to bikeshed it here. Give the Makefile a try, you'll be surprised how nice it makes the dev process and how it actually documents what is done at each step.

Comment (haadcode) [inline Makefile]: That said, I'm highly in favor of adding Makefiles. @jbenet's been going on about them for a long time and I didn't pay attention until very recently and he's totally right about their value. They just make it so much easier to get started than decoding what's in a specific package.json file.

File: src/index.js (hunk 5 above)
Comment (dignifiedquire) [inline src/index.js]: how is direct assignment less clearer than using `Object.assign`?

Comment (dignifiedquire) [inline src/index.js]: For me it's the exact opposite, also why not just use a single call to `Object.assign`? I had to go through the whole code checking if this wasn't overriding something because that is what `Object.assign` is usually used for, merging objects with possible conflicts

Comment (dignifiedquire) [inline src/index.js]: Though to be honest, I think this would be a much better interface, than assigning props to a function

```js
exports.download = download
exports.platforms = support.platforms
exports.versions = support.versions
exports.archs = support.archs
```

Comment (haadcode) [inline src/index.js]: I refactored the public api. I still think the uppercased properties are nicer, so keeping them as such. We need to have the download function as our module.export in order to not break the API for users (backwards compatibility).

File: src/bin.js (hunk 2 above)
Comment (haadcode) [inline src/bin.js]: Oh, I can't :/

File: src/index.js (hunk 5 above)
Comment (haadcode) [inline src/index.js]: Take that back, that doesn't compile :)

Comment (haadcode) [inline src/index.js]: You know what, let's take the Versions, Platforms and Archs out of the public api. They're not really what this module *needs* to do, so let's keep it the way it was. If we need such functionality in the future, we can add something like `getAllPackages()` or smth that then returns the data.

This way the code is the clearest out of all options and we don't add a need to support a larger api surface :)

Comment (haadcode) [inline src/index.js]: Changed

File: Makefile (hunk 6 above)
Comment (jbenet) [inline Makefile]: > I am 👎 on adding makefiles, they are not cross platform and don't add much value compared to just npm run-script

Wrong. They add an enormous amount of value for users of tools cross language. The point of `make` is that -- no matter who wrote something and what they use daily -- a `Makefile` is a language-agnostic tool, so anyone can jump in and understand what to do. I'm not suggesting you use `make` to do the heavy lifting. I'm suggesting the use of `Makefiles` in places like Orbit, or orbit-db, where people will download the repos and want to do something, but (a) don't know npm well. (b) don't know how to invoke things. (c) dont know _which_ things to invoke. `make` is a _standard_ way to do it.

also, i dont know what the current distribution of people with Windows, unix tooling, and no make program. It used to be quite small. maybe someone should compile `make` with emscripten and publish it to npm.

Comment (jbenet) [inline Makefile]: Also, i'm not advocating the horrendous autotools or Cmake makefiles that are so confusing and hard to figure out. I'm advocating basic things like this:
- https://github.com/jbenet/macos-ssh-proxy/blob/master/Makefile
- https://github.com/jbenet/go-random-files/blob/master/Makefile
- https://github.com/jbenet/gob-js-test/blob/master/Makefile (particularly see the deps part)

obviously none of this is perfect, and i dont use them in all my repos, but i've learned that it's a great way to signal "what this thing should do" and "how to use this thing" in a language everyone can understand.

Comment (haadcode) [inline Makefile]: 👍  to all what @jbenet said: keep 'em simple, use them to document *how* to get a repo up an running (*which* things to run is the most important description here), not for heavy lifting (npm is fine for that in JS, other langs have their own tools).

@jbenet also: cool idea to add command help to `make all`.

Comment (haadcode) [inline Makefile]: What I've found hugely valuable with Makefiles is with the "immutability". `make clean` makes sure to reset the repo to its original state and testing becomes much more reliable when `node_modules/` gets wiped and re-installed from clean slate. This was the reason I started adding them to my repos. They give a really simple and clean set of instructions to run on a repo to make sure everything is working: `make clean && make` and all the expected things should've been run & built (dist builds, tests, installing deps, etc).

I would recommend the following approach: try it out, add it to the repos you work on and see if it is valuable for you. It might not make sense to add them *everywhere* but for something like this repo, or js-ipfs-api, I think they work great and give a really clean way to develop. YMMV. And I would imagine for newcomers it's easier to check which commands to run from a Makefile as opposed to package.json.

As mentioned above, Windows is the black sheep in this case too, so let's make sure any repo doesn't *depend* on Makefiles (ie. it's not the primary location to define the development tooling), they're additional info.

Comment (dignifiedquire) [inline Makefile]: You can get the same thing, cross platform with

```json
"scripts": {
  "clean": "rimraf node_modules",
  "test": "mytestcommand"
}
```

and `npm run clean && npm i && npm test`. If you want to have it even shorter, just add an alias to scripts 

```json
"make": "npm run clean && npm install && npm test"
```

Comment (haadcode) [inline Makefile]: Hey, I removed the Makefile from this PR as to not keep us back from merging this. I separated it to a new PR in https://github.com/ipfs/go-ipfs-dep/pull/15. 

This was a good discussion and at least the link to it should be kept safe for future references :)

**Comment by haadcode** [comment]
Ok. Everything addressed. Can I get a CR and LGTM if all good to merge?

File: src/bin.js [hunk 7]
```diff
@@ -2,4 +2,9 @@
 
 'use strict'
 
-require('./')()
+// First param is the target version
+// Second param is the target platform
+// Third param is the target architecture
+require('./')(process.argv[2], process.argv[3], process.argv[4])
+  .then(() => process.exit(0))
+  .catch((e) => process.exit(1))
```
Comment (dignifiedquire) [inline src/bin.js]: Wouldn't it be good to print the error and not just exit?

File: package.json [hunk 8]
```diff
@@ -2,6 +2,9 @@
   "name": "go-ipfs-dep",
   "version": "0.4.4",
   "description": "Install the latest go-ipfs binary",
+  "go-ipfs": {
+    "version": "v0.4.4"
+  },
```
Comment (dignifiedquire) [inline package.json]: Needs engines set to something like this: https://github.com/ipfs/js-ipfs-api/blob/master/package.json#L51-L54  as it is now using things like arrow functions, which are not available earlier

**Comment by dignifiedquire** [comment]
@haadcode two last things, other than it looks good to me if CI is happy

File: src/bin.js (hunk 7 above)
Comment (haadcode) [inline src/bin.js]: The errors (all output) are in index.js / error(). They get output, see the tests.

File: package.json (hunk 8 above)
Comment (haadcode) [inline package.json]: Fixed.

File: src/bin.js (hunk 7 above)
Comment (haadcode) [inline src/bin.js]: But you're right, the actual download function doesn't need to handle the output and it can be moved to bin.js (cli). Makes more sense in terms of responsibilities. Will change that.

Comment (haadcode) [inline src/bin.js]: Changed.

**Comment by dignifiedquire** [comment]
LGTM

File: .travis.yml [hunk 9]
```diff
@@ -10,16 +10,12 @@ before_install:
   - npm install -g npm
 
 script:
-  - npm run lint
```
Comment (diasdavid) [inline .travis.yml]: Why stop linting?

File: .travis.yml [hunk 10]
```diff
@@ -10,16 +10,12 @@ before_install:
...
   - npm test
 
 before_script:
   - export DISPLAY=:99.0
   - sh -e /etc/init.d/xvfb start
 
-after_success:
-  - npm run coverage-publish
```
Comment (diasdavid) [inline .travis.yml]: Why stop publishing coverage?

File: package.json [hunk 11]
```diff
@@ -2,11 +2,16 @@
...
   "main": "src/index.js",
+  "engines": {
+    "node": ">=4.0.0",
+    "npm": ">=3.0.0"
+  },
   "scripts": {
     "install": "node src/bin.js",
-    "lint": "aegir-lint",
-    "coverage": "aegir-coverage",
```
Comment (diasdavid) [inline package.json]: ?

File: src/check-support.js [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+    throw new Error(`Version '${version}' not available`)
+
+  return true
+}
+
+// Public API
+module.exports = {
+  Versions: supportedVersions,
+  Platforms: supportedPlatforms,
+  Archs: supportedArchs,
```
Comment (diasdavid) [inline src/check-support.js]: Do these have to be uppercase?

File: test/index.js [hunk 13]
```diff
@@ -1,20 +1,145 @@
...
-var fs = require('fs')
-var path = require('path')
-var rimraf = require('rimraf')
+const test = require('tape')
+const fs = require('fs')
+const path = require('path')
+const rimraf = require('rimraf')
+const goenv = require('go-platform')
+const pkg = require('./../package.json')
+const Download = require('../src')
```
Comment (diasdavid) [inline test/index.js]: Is this a class? If not, can we use lowercase?

File: package.json (hunk 8 above)
Comment (haadcode) [inline package.json]: You can use one or the other. If you don't specify the "go-ipfs.version" field, then the package version is used. This is to enable us to release a) multiple version of the same go-ipfs version (eg. 0.4.5-1) and b) enable us to release dev versions of go-ipfs.

File: .travis.yml (hunk 10 above)
Comment (haadcode) [inline .travis.yml]: Same as for linting.

File: package.json (hunk 11 above)
Comment (haadcode) [inline package.json]: See above.

File: src/check-support.js (hunk 12 above)
Comment (haadcode) [inline src/check-support.js]: I prefer the upper case notation here.

File: test/index.js (hunk 13 above)
Comment (haadcode) [inline test/index.js]: Nope, not a class, just a function. See couple of lines below in the test file why that is wrapped like that.

**Comment by haadcode** [comment]
@diasdavid Thanks for the review. I thought I had updated README in this PR but obviously didn't, so good catch (it was dropped somewhere down the line of merging and rebasing). Added it to the PR.

**Comment by haadcode** [comment]
@diasdavid as for npm-go-ipfs, it only uses go-ipfs-dep to figure out a path (see https://github.com/ipfs/npm-go-ipfs/blob/master/link-ipfs.js#L9), so this PR will not effect it.

File: README.md [hunk 14]
```diff
@@ -17,8 +17,63 @@ npm install go-ipfs --save
...
 **Warning**: The binary gets put in the `go-ipfs` folder inside the module folder.
+
+### Which go-ipfs version this package downloads?
+
+Can be specified in `package.json` with a field `go-ipfs.version`, eg:
+
+```json
+"go-ipfs": {
+  "version": "v0.4.5-pre1"
+},
```
Comment (diasdavid) [inline README.md]: super sweeeeeeeeet! ❤️❤️❤️❤️❤️❤️

File: README.md [hunk 15]
```diff
@@ -17,8 +17,63 @@ npm install go-ipfs --save
...
+
+### Using local IPFS daemon as the package download url
+Can be specified in `package.json` with a field `go-ipfs.version`, eg:
+
+```json
+"go-ipfs": {
+  "version": "v0.4.5-pre1",
+  "distUrl": "http://localhost:8080/ipfs/QmSoNtqW22htkg9mtHWNBvZLUEmqfq8su7957meS1iQfeL"
+},
+```
```
Comment (diasdavid) [inline README.md]: ❤️❤️❤️❤️❤️❤️❤️

Comment (haadcode) [inline README.md]: I'm sure you got it, but just to make clear, this (the distUrl param) is to make it possible to a) test unreleased dist sites (so that whoever publishes one can be sure it doesn't break the whole of JS' go-ipfs ecosystem) and b) can test and use this via IPFS gateway (which makes it *extremely* fast) or locally.

Comment (diasdavid) [inline README.md]: Yeah, that is way I was so excited :D
